### ISSUE TITLE
Update SwiftyJSON dependency for compatibility with Swift 3.0.1?

### ISSUE THREAD

**Issue by elimist** [MLSDev/TRON#20]
Hello,

I tried using TRON, but my project will not compile, saying that `SwiftyJSON` is not compatible with Swift 3.0.1. Also, is it possible to install TRON core through Carthage without the SwiftyJSON dependency like with Cocoapods?

**Comment by DenTelezhkin** [comment]
Hey. You probably need to recompile frameworks, since Swift 3 is not ABI-stable, and frameworks compiled with Swift 3.0 compiler is not compatible with Swift 3.0.1 compiler.
We use TRON with SwiftyJSON on both Swift 3.0 and Swift 3.0.1 without issues. You just need to recompile from code every time.

As for subspecs, Carthage does not have this concept unfortunately. So with Carthage you get everything - or nothing. There was an idea to split subspecs to different targets but the project setup seems too complicated to do that, and we opted not to do that right now.

We are looking forward to splitting subspecs to parts with Swift Package Manager(or Carthage, if it chooses to implement such feature), and compiling binary frameworks in Swift 4, but we are not there yet, unfortunately.

**Comment by elimist** [comment]
Hey, thanks for the reply. In regards to Carthage and SwiftyJSON, I think the way Moya does it is to generate multiple frameworks and then users can just link whichever framework they need:

From the Readme:

> Carthage users can point to this repository and use whichever generated framework they'd like, Moya, RxMoya, or ReactiveMoya. The full Moya framework is bundled in each of those frameworks; importing more than one framework in a single file will result in ambiguous lookups at compile time.

I don't know if this is some thing you would consider doing for this project, but to me it would make sense since TRON was designed to be modular and pluggable.

**Comment by DenTelezhkin** [comment]
I was trying to simplify repo setup by first simplifying dependency tree
https://github.com/Alamofire/Alamofire/pull/1064

But unfortunately Alamofire maintainers did not go with proposed xcodeproj structure. So, right now, TRON.xcodeproj has 7 targets - 4 targets for frameworks on macOS,iOS, tvOS and watchOS, and 3 targets for unit tests - on macOS, iOS, tvOS. Moya setup is even crazier - they have 12 framework targets and 3 test targets, which as it seems does not cover everything.

For me personally splitting project into parts would be maintanable if we had simply TRON.framework for all platforms, and TRONTests. Then adding TRON/RxSwift and TRON/SwiftyJSON targets would be perfectly acceptable, and i think we would even add more like TRON/Unbox or TRON/mapper. But right now each added target is basically multiplied by 4 - for each platform, or even by 7, if you add tests on top of it. Compare this to CocoaPods, which does not need any of this, and splits into subspecs even though all these files exist in single project.

So in short - i really wanted to do this, but the overhead to do this just for Carthage is insane, so it was put into ice-box for now.

**Comment by elimist** [comment]
I see what you mean. I'm really not impressed with Carthage's insistence on not allowing subspecs in the name of "preserving small frameworks"... I myself favour composability and modularity over the batteries-included approach, but I don't think that's something a package manager should dictate. I can see that it would be a huge amount of work for frameworks targeting multiple platforms.

I just started using Carthage a few weeks ago since it seemed like the 'modern' way of doing things, but now I think i'm just going to return to using Cocoapods.

**Comment by DenTelezhkin** [comment]
Yeah, anyway for Swift 3 CocoaPods seem to be really great, because until Swift 4 we are supposed to be building frameworks from source to avoid binary-incompatibility, and CocoaPods is the only package manager that builds from source =) I hope with Swift 4 situation will improve and we'll finally use prebuilt frameworks(and hopefully subframeworks as well).
